Tests/FuzzilliTests/AnalyzerTest.swift

```diff
@@ -31,15 +31,127 @@ class AnalyzerTests: XCTestCase {
+            }
+            XCTAssertEqual(b.context, [.script, .function, .asyncFunction])
+            b.await(value: v3)
+            b.defineAsyncGeneratorFunction(withSignature: FunctionSignature(withParameterCount: 2)) { _ in
+                XCTAssertEqual(b.context, [.script, .function, .asyncFunction, .generatorFunction])
+            }
+            XCTAssertEqual(b.context, [.script, .function, .asyncFunction])
+        }
+
+        let _ = b.finalize()
+    }
+
+    func testNestedWithStatements() {
+        let fuzzer = makeMockFuzzer()
+        let b = fuzzer.makeBuilder()
+
+        XCTAssertEqual(b.context, .script)
+        let obj = b.loadString("HelloWorld")
+        b.with(obj) {
+            XCTAssertEqual(b.context, [.script, .with])
+            b.definePlainFunction(withSignature: FunctionSignature(withParameterCount: 3)) { _ in
+                XCTAssertEqual(b.context, [.script, .function])
+                b.with(obj) {
+                    XCTAssertEqual(b.context, [.script, .function, .with])
+                }
+            }
+            XCTAssertEqual(b.context, [.script, .with])
+            b.loadFromScope(id: b.genPropertyNameForRead())
+        }
+
+        let _ = b.finalize()
+    }
+
+    func testClassDefinitions() {
+        let fuzzer = makeMockFuzzer()
+        let b = fuzzer.makeBuilder()
+
+        XCTAssertEqual(b.context, .script)
+        let superclass = b.defineClass() { cls in
+            cls.defineConstructor(withParameters: [.integer]) { params in
+                XCTAssertEqual(b.context, [.script, .classDefinition, .function])
+                let loopVar1 = b.loadInt(0)
+                b.doWhileLoop(loopVar1, .lessThan, b.loadInt(42)) {
+                    XCTAssertEqual(b.context, [.script, .classDefinition, .function, .loop])
+                }
+                XCTAssertEqual(b.context, [.script, .classDefinition, .function])
+            }
+        }
+        XCTAssertEqual(b.context, .script)
+
+        b.defineClass(withSuperclass: superclass) { cls in
+            cls.defineConstructor(withParameters: [.string]) { _ in
+                XCTAssertEqual(b.context, [.script, .classDefinition, .function])
+                let v0 = b.loadInt(42)
+                let v1 = b.createObject(with: ["foo": v0])
+                b.callSuperConstructor(withArgs: [v1])
+            }
+            cls.defineMethod("classMethod", withSignature: FunctionSignature(withParameterCount: 2, hasRestParam: false)) { _ in
+                XCTAssertEqual(b.context, [.script, .classDefinition, .function])
+                b.defineAsyncFunction(withSignature: FunctionSignature(withParameterCount: 2)) { _ in
+                    XCTAssertEqual(b.context, [.script, .function, .asyncFunction])
+                }
+                XCTAssertEqual(b.context, [.script, .classDefinition, .function])
+            }
+        }
+        XCTAssertEqual(b.context, .script)
+
+        let _ = b.finalize()
+    }
```

Comment:
Maybe also add a test for `CodeStrings`, since you (correctly!) changed them to not propagate the surrounding context.
Also maybe add one testcase, `testTrivialBlocks` or so, that tests various blocks that don't create new contexts, such as block statements, if-else, try-catch. (just open some context like `.function` + `.loop`, then create each of these blocks inside it and verify the context inside is equal to the one outside)

Reply:
testcases added